modules/common/src/main/Markdown.scala

```diff
@@ -82,8 +99,94 @@ final class Markdown(
 
 object Markdown {
 
-  private val imageRegex = """!\[[^\]]*\]\((.*?)\s*("(?:.*[^"])")?\s*\)""".r
-
-  def imageUrls(markdown: String): List[String] =
-    imageRegex.findAllIn(markdown).matchData.map(_ group 1).toList
+  private val rel = "nofollow noopener noreferrer"
+
+  private object WhitelistedImageExtension extends HtmlRenderer.HtmlRendererExtension {
+    override def rendererOptions(options: MutableDataHolder) = ()
+    override def extend(htmlRendererBuilder: HtmlRenderer.Builder, rendererType: String) =
+      htmlRendererBuilder
+        .nodeRendererFactory(new NodeRendererFactory {
+          override def apply(options: DataHolder) = WhitelistedImageNodeRenderer
+        })
+        .unit
+  }
+  private object WhitelistedImageNodeRenderer extends NodeRenderer {
+    override def getNodeRenderingHandlers() =
+      new java.util.HashSet(
+        Arrays.asList(
+          new NodeRenderingHandler(classOf[Image], render)
+        )
+      )
+
+    private val whitelist =
+      List(
+        "imgur.com",
+        "giphy.com",
+        "wikimedia.org",
+        "creativecommons.org",
+        "pexels.com",
+        "piqsels.com",
+        "freeimages.com",
+        "unsplash.com",
+        "pixabay.com",
+        "githubusercontent.com",
+        "googleusercontent.com",
+        "i.ibb.co",
+        "i.postimg.cc",
+        "xkcd.com",
+        "lichess1.org"
+      )
+    private def whitelistedSrc(src: String): Option[String] =
+      for {
+        url <- Try(URL.parse(src)).toOption
+        if url.scheme == "http" || url.scheme == "https"
+        host <- Option(url.host).map(_.toString)
+        if whitelist.exists(h => host == h || host.endsWith(s".$h"))
+      } yield url.toString
+
+    private def render(node: Image, context: NodeRendererContext, html: HtmlWriter): Unit =
+      // Based on implementation in CoreNodeRenderer.
+      if (context.isDoNotRenderLinks || CoreNodeRenderer.isSuppressedLinkPrefix(node.getUrl(), context))
+        context.renderChildren(node)
+      else {
+        val resolvedLink = context.resolveLink(LinkType.IMAGE, node.getUrl().unescape(), null, null)
+        val url          = resolvedLink.getUrl()
+        val altText      = new TextCollectingVisitor().collectAndGetText(node)
+        whitelistedSrc(url) match {
+          case Some(src) =>
+            html
+              .srcPos(node.getChars())
+              .attr("src", src)
+              .attr("alt", altText)
+              .attr(resolvedLink.getNonNullAttributes())
+              .withAttr(resolvedLink)
+              .tagVoid("img")
+          case None =>
```

Comment:
Main downside: At this point it's too late to report a form validation error, so we render a link instead of embedding the image. But this way there's no opportunity to explain what's going on and show `Allowed domains: ...`.